src/java/org/apache/cassandra/utils/Overlaps.java

```diff
@@ -123,44 +125,138 @@ public static <E, B> List<B> assignOverlapsIntoBuckets(int threshold,
-            if (maxOverlap < threshold)
+            final int i = bySize.next();
+            if (used.get(i))
                 continue;
+
+            Set<E> bucket = overlaps.get(i);
+            if (bucket.size() < threshold)
+                break;  // no more buckets will be above threshold
+            used.set(i);
+
+            Set<E> allOverlapping = bucket;
+            int j = i - 1;
+            int k = i + 1;
             int startIndex = i;
             int endIndex = i + 1;
-
-            if (inclusionMethod != InclusionMethod.NONE)
+            // expand to include neighbors that intersect with current bucket
+            if (inclusionMethod == InclusionMethod.SINGLE)
             {
-                Set<E> allOverlapping = new HashSet<>(bucket);
-                Set<E> overlapTarget = inclusionMethod == InclusionMethod.TRANSITIVE
-                                       ? allOverlapping
-                                       : bucket;
-                int j;
-                for (j = i - 1; j >= lastEnd; --j)
+                // expand the bucket to include all overlapping sets
+                allOverlapping = new HashSet<>(bucket);
+                Set<E> overlapTarget = bucket;
+                for (; j >= 0 && !used.get(j); --j)
                 {
                     Set<E> next = overlaps.get(j);
                     if (!setsIntersect(next, overlapTarget))
                         break;
                     allOverlapping.addAll(next);
+                    used.set(j);
                 }
                 startIndex = j + 1;
-                for (j = i + 1; j < regionCount; ++j)
+                for (; k < regionCount && !used.get(k); ++k)
                 {
-                    Set<E> next = overlaps.get(j);
+                    Set<E> next = overlaps.get(k);
                     if (!setsIntersect(next, overlapTarget))
                         break;
                     allOverlapping.addAll(next);
+                    used.set(k);
                 }
-                i = j - 1;
-                endIndex = j;
+                endIndex = k;
+            }
+            // Now mark all overlapping with the extended as used
+            Set<E> overlapTarget = allOverlapping;
+            for (; j >= 0 && !used.get(j); --j)
+            {
+                Set<E> next = overlaps.get(j);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                used.set(j);
+                unselectedHandler.accept(next);
             }
+            for (; k < regionCount && !used.get(k); ++k)
+            {
+                Set<E> next = overlaps.get(k);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                used.set(k);
+                unselectedHandler.accept(next);
+            }
+            buckets.add(bucketer.makeBucket(overlaps, startIndex, endIndex));
+        }
+        for (int i = 0; i < regionCount; ++i)
```

Comment:
Nit: Apologies for forgetting this in the code I sent, but these lists can be of non-trivial size (10s of thousands in the extreme), so it makes sense to rewrite this to use `used.nextClearBit()` as
```
        for (int i = used.nextClearBit(0); i < regionCount; i = used.nextClearBit(i + 1))
            unselectedHandler.accept(overlaps.get(i));
```

Reply:
updated